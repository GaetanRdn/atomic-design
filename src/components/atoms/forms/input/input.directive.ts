import { Directive, EventEmitter, HostBinding, HostListener, Input, OnDestroy, Output } from '@angular/core';
import { CoerceBoolean } from '../../../core/common/coerce-boolean-inputs.decorator';

@Directive({
  selector: 'input[adrInput]',
  host: {
    class: 'adr-input',
    '[class.adr-focused]': 'focused || null'
  }
})
export class InputDirective implements OnDestroy {
  @HostBinding('attr.value')
  @Input()
  public value: any;

  @Input()
  @CoerceBoolean()
  public readonly: boolean = false;

  @Input()
  @CoerceBoolean()
  public disabled: boolean = false;

  @Output()
  public valueChange: EventEmitter<any> = new EventEmitter<any>();

  private _focused: boolean = false;

  get focused(): boolean {
    return this._focused;
  }

  @HostListener('input', ['$event.target.value'])
  public onInput(value: any): void {
    if (!this.readonly && !this.disabled) {
      this.valueChange.emit(value);
    }
  }

  @HostListener('focus', ['true'])
  @HostListener('blur', ['false'])
  public onToggleFocus(focused: boolean): void {
    if (!this.readonly && !this.disabled) {
      this._focused = focused;
    }
  }

  public ngOnDestroy(): void {
    this.valueChange.unsubscribe();
  }
}
