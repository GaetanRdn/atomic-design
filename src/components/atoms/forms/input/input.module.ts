import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InputDirective } from './input.directive';



@NgModule({
  declarations: [
    InputDirective
  ],
  imports: [
    CommonModule
  ],
  exports: [
    InputDirective
  ]
})
export class InputModule { }
