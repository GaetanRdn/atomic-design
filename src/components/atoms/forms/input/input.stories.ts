import { Component, EventEmitter, Input, Output } from '@angular/core';
import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/angular';
import { InputDirective } from 'src/components/atoms/forms/input/input.directive';

@Component({
  selector: 'adr-input-dummy',
  template: ` <input
    adrInput
    [disabled]="disabled"
    [readonly]="readonly"
    [value]="value"
    (valueChange)="valueChange.emit($event)"
    placeholder="default input"
  />`,
})
class InputDummyComponent {
  @Input()
  public value: any;

  @Input()
  public readonly: boolean = false;

  @Input()
  public disabled: boolean = false;

  @Output()
  public valueChange: EventEmitter<any> = new EventEmitter<any>();
}

export default {
  title: 'atoms/forms/input',
  argTypes: {
    value: { control: { type: 'text' } },
  },
  component: InputDummyComponent,
  parameters: {
    jest: ['input.directive.spec.ts'],
  },
} as Meta;

const BasicTemplate: Story<InputDummyComponent> = (args) => ({
  props: {
    ...args,
    valueChange: action('log'),
  },
  moduleMetadata: { declarations: [InputDirective] },
});

export const Default = BasicTemplate.bind({});
