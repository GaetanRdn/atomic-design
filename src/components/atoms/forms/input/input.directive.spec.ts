import { Component, DebugElement } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { TemplateLookup } from '../../../core/tests/template-lookup';
import { InputDirective } from './input.directive';

describe('InputDirective', () => {

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [InputDirective, BasicComponent, ReadonlyComponent, DisabledComponent]
    });
  });

  describe('Basic input', () => {
    let templateLookup: TemplateLookup<BasicComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(TestBed.createComponent(BasicComponent));
    });

    test('Check default state', () => {
      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test.each(['value1', 'value2'])('should set value %s', (value: string) => {
      // GIVEN
      templateLookup.hostComponent.value = value;

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.query<HTMLInputElement>('input').value).toEqual(value);
    });

    test.each(['value1', 'value2'])('should update parent value on change, expect %s', (value: string) => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      triggerInputValue(templateLookup, value);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toEqual(value);
    });

    test('check class on focus', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      let input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when focused then check class on blur', () => {
      // GIVEN
      let input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // WHEN
      input.triggerEventHandler('blur', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Readonly input', () => {
    let templateLookup: TemplateLookup<ReadonlyComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(TestBed.createComponent(ReadonlyComponent));

      templateLookup.detectChanges();
    });

    test('do not emit change', () => {
      // WHEN
      triggerInputValue(templateLookup, 'value');
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test('when focus then check dont have focused class', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      let input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Disabled input', () => {
    let templateLookup: TemplateLookup<DisabledComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(TestBed.createComponent(DisabledComponent));

      templateLookup.detectChanges();
    });

    test('do not emit change', () => {
      // WHEN
      triggerInputValue(templateLookup, 'value');
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test('when focus then check dont have focused class', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      let input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

const triggerInputValue = (templateLookup: TemplateLookup<unknown>, value: any): void => {
  let input: DebugElement = templateLookup.get('input');
  input.nativeElement.value = value;
  input.triggerEventHandler('input', { target: input.nativeElement });
};

@Component({
  template: `<input adrInput [(value)]="value" />`
})
class BasicComponent {
  public value!: string;
}

@Component({
  template: `<input adrInput readonly [(value)]="value" />`
})
class ReadonlyComponent {
  public value = null;
}

@Component({
  template: `<input adrInput disabled [(value)]="value" />`
})
class DisabledComponent {
  public value = null;
}
