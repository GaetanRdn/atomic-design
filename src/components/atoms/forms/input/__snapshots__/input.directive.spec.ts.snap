// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDirective Basic input Check default state 1`] = `
<input
  adrinput=""
  class="adr-input"
/>
`;

exports[`InputDirective Basic input check class on focus 1`] = `
<input
  adrinput=""
  class="adr-input adr-focused"
/>
`;

exports[`InputDirective Basic input should set value value1 1`] = `
<input
  adrinput=""
  class="adr-input"
  value="value1"
/>
`;

exports[`InputDirective Basic input should set value value2 1`] = `
<input
  adrinput=""
  class="adr-input"
  value="value2"
/>
`;

exports[`InputDirective Basic input should update parent value on change, expect value1 1`] = `
<input
  adrinput=""
  class="adr-input"
  value="value1"
/>
`;

exports[`InputDirective Basic input should update parent value on change, expect value2 1`] = `
<input
  adrinput=""
  class="adr-input"
  value="value2"
/>
`;

exports[`InputDirective Basic input when focused then check class on blur 1`] = `
<input
  adrinput=""
  class="adr-input"
/>
`;

exports[`InputDirective Disabled input do not emit change 1`] = `
<input
  adrinput=""
  class="adr-input"
  disabled=""
/>
`;

exports[`InputDirective Disabled input when focus then check dont have focused class 1`] = `
<input
  adrinput=""
  class="adr-input"
  disabled=""
/>
`;

exports[`InputDirective Readonly input do not emit change 1`] = `
<input
  adrinput=""
  class="adr-input"
  readonly=""
/>
`;

exports[`InputDirective Readonly input when focus then check dont have focused class 1`] = `
<input
  adrinput=""
  class="adr-input"
  readonly=""
/>
`;
